tests/ui/typeck/closure-ty-mismatch-issue-128561.stderr

```diff
@@ -0,0 +1,26 @@
+error[E0308]: mismatched types
+  --> $DIR/closure-ty-mismatch-issue-128561.rs:2:32
+   |
+LL |     b"abc".iter().for_each(|x| x);
+   |                                ^
+   |                                |
+   |                                expected `()`, found `&u8`
+   |                                help: consider ignore the value here: `_ =`
+
+error[E0308]: mismatched types
+  --> $DIR/closure-ty-mismatch-issue-128561.rs:4:32
+   |
+LL |     b"abc".iter().for_each(|x| dbg!(x));
+   |                                ^^^^^^^ expected `()`, found `&u8`
+   |
+   = note: this error originates in the macro `dbg` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Comment:
I notice that the suggestion *didn't* trigger for the macro case, which is what the original report wanted. I think that in order to sidestep the issues of the macro not being visible in the HIR, to get the span by looking at the closure's tail expression (instead of `expression`, get the parent node and get the span from there).

Reply:
I didn't notice in the macro case the parent type check here is not closure:

https://github.com/rust-lang/rust/pull/135562/files#diff-4931c060d9b3dc4c4a95c29607277bce4b456a157c7a72401c3f0441d03b76b3L1896-L1899

I changed the code to avoid suggesting for macro case explicitly.


Seems we need to get the parent hir recursively up to 4 levels in macro case to get the closure body, I tried with some debug code like:

https://github.com/chenyukang/rust/blob/3e380a8e3b2d7b9416ac7dedce6b5b055ca9a9a3/compiler/rustc_middle/src/hir/map/mod.rs#L169-L183

https://github.com/chenyukang/rust/blob/3e380a8e3b2d7b9416ac7dedce6b5b055ca9a9a3/compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs#L813-L814

Reply:
There are two options that might or might not work: from the `expression`, climb up its parents until you get a node with a span with root context (hence, not pointing at a macro). That runs the risk of giving you the node of an expression still surrounded by a macro, which isn't great. I think this suggestion is only relevant for the tail expression of closures, right? The second option would be to get the parent closure by looping on getting the parent until you hit one (with `get_closure_node`, as you point out), and then look down from there. A macro that uses closures internally can still mess your suggestion up though, so you likely still need a check for "what's this span's context", regardless of what you do :-/